src/core/qgsvectorlayer.cpp

```diff
@@ -2387,6 +2387,19 @@ long QgsVectorLayer::featureCount() const
          ( mEditBuffer ? mEditBuffer->mAddedFeatures.size() - mEditBuffer->mDeletedFeatureIds.size() : 0 );
 }
 
+long QgsVectorLayer::featureCount( QgsFeatureRequest request )
+{
+  QgsFeatureIterator fit = getFeatures( request );
```

Comment:
Hi, QgsFeatureRequest could be created fetching minimal data, it is faster.

See:
https://github.com/rldhont/Quantum-GIS/blob/featureCount_with_featureRequest/src/core/qgsvectorlayer.cpp#L500